plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -165,7 +177,8 @@ export class PersonState {
     }
 
     private async updatePersonProperties(): Promise<Person> {
-        const personFound = await this.db.fetchPerson(this.teamId, this.distinctId)
+        // Note: We always fetch the person anew to avoid property-related races.
+        const personFound = this.person || (await this.db.fetchPerson(this.teamId, this.distinctId))
```

Comment:
also what's the impact of using this cached person here? we will be more subject to races, no? or what's the analysis that led to the decision of using the cache here? in order ingestion? (even though races are possible with identify)

Reply:
> I have a counter-question: Why would we want to refetch persons multiple times in this pipeline? I can see it being no less prone to races and leads directly worse performance. 

Agreed it's not good for perf nor Postgres load.

> The only reason we were consistently fetching these persons didn't have anything to do with correctness but just a side-effect of poorly tested code

I wish this was true, unfortunately it is tied to correctness here due to the way we do props updates, but we might say yolo and ignore the potential races and get the perf win (fine by me, we should re-design props updating, but I don't have the proposal written up yet).

Here's an example of what can happen now:

Let's say we have two executions running in parallel: 
1. A - which does an alias
2. B which does a random property update. 
 
Note that due to the fact that A does an alias it's possible that these are in different partitions, which means could be processed in parallel even after our recent changes to break batches up by distinctId.

Imagine the following execution flow in time: 
1. A reached here `const personFound = this.person` and uses the person they merged from the beginning. 
2. B reaches here, fetches the person (if not with `for update` we're prone to this problem here anyway)
5. B updates some properties & completes execution
6. A computes the props update agains the person from step (1)/earlier & overrides changes B made (as we just dump the whole json that happens even if different props were changed).

Reply:
@tiina303 the exact same race condition still exists in the previous version as the alias could happen between fetchPerson in `updateProperties` and the actual updating.

By re-fetching I posit we were not actually solving any races - just making them harder to detect.

Reply:
Would having a metric like [this](https://github.com/PostHog/posthog/pull/10423) help us feel more comfortable releasing this @tiina303?

Reply:
I tried to say twice in my comment that it's good to be merged & didn't request changes. So let me repeat: it's good to go in once Yakko or Michael approved. Sorry if that wasn't clear, I'll be more careful about my wording "yolo" in the future. I did want to call out there this makes things a bit worse & provided an example. The reason I commented in the first place was so it would be faster for us to find / troubleshoot if after merging we see problems.